File: Programmers/문제풀이_구/Level 2/Level2_연속된부분수열의합.py
sequence = [1, 1, 1, 2, 3, 4, 5]
k = 5

# 1트(시간초과), while을 통해 popleft()하는 과정이 시간을 많이 잡아먹는것 같음
from collections import deque
def solution(sequence, k):
    answer = []

    start = -1
    end = -1
    min_len = 987654321
    q = deque()
    for i in range(len(sequence)):
        if k < sequence[i]:
            break

        q.append(sequence[i])
        if start == -1:
            start = i

        totalSequence = sum(q)
        while totalSequence > k:
            q.popleft()
            start += 1
            totalSequence = sum(q)

        if totalSequence == k:
            end = start + len(q) - 1

            if len(q) < min_len:
                min_len = len(q)
                answer.clear()
                answer.append(start)
                answer.append(end)
                if min_len == 1:
                    return answer            

        print(str(q) + " / {} = {}".format(answer, min_len) )

    return answer


# 2트(성공), deque를 제거하고 직접 위치를 옮겨주는 방식으로 수정후 성공함
def solution2(sequence, k):
    start = 0
    end = 0
    result_start = -1
    result_end = -1
    min_len = 987654321
    total = sequence[0]

    while True:
        if total <= k:
            if total == k:
                if min_len > end - start + 1:
                    min_len = end - start + 1
                    result_start = start
                    result_end = end
            end += 1
            if end >= len(sequence):
                break
            total += sequence[end]
        else:
            start += 1
            total -= sequence[start - 1]

    return [result_start, result_end]


print(solution2(sequence, k))

File: Programmers/Level 2/Level2_연속된부분수열의합.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<int> solution(vector<int> sequence, int k) {
    vector<int> answer;

    int start = 0;
    int end = 0;
    int min_start = -1;
    int min_end = -1;
    int min_len = INT_MAX;
    int total = sequence[0];

    while (start < sequence.size() && end < sequence.size())
    {
        // 기존보다 길이가 짧다면
        if (total == k && (end - start + 1 < min_len))
        {
            min_start = start;
            min_end = end;
            min_len = end - start + 1;
        }

        if (total < k)
        {
            end += 1;
            if (end >= sequence.size())
                break;
            total += sequence[end];
        }
        else
        {
            start += 1;
            if (start >= sequence.size())
                break;
            total -= sequence[start - 1];
        }
    }

    answer.push_back(min_start);
    answer.push_back(min_end);
    return answer;
}


// 1차 시도 : 성공
// 이전에 파이썬으로 풀었던 코드가 있어서 비교해보니 예나 지금이나 생각하는 방식은 동일했다.